XRPLib/imu.py

```diff
@@ -56,10 +59,24 @@ def __init__(self, scl_pin: int, sda_pin: int, addr):
         self._power = True
         self._power_a = 0x10
         self._power_g = 0x10
-        # ODR_XL=1 FS_XL=0
-        self._setreg(LSM6DSO_CTRL1_XL, 0x10)
-        # ODR_G=1 FS_125=1
-        self._setreg(LSM6DSO_CTRL2_G, 0x12)
+
+        # Check WHO_AM_I register to verify IMU is connected
+        foo = self._getreg(LSM6DSO_WHO_AM_I)
+        if foo != 0x6C:
+            # Getting here indicates sensor isn't connected
+            # TODO - do somehting intelligent here
```

Comment:
This probably ties into the other comment I made regarding the best way to architect the library. In general, what should the library do when an error occurs? I would argue that throwing exceptions would be best when the problem can't be easily solved. We could even define custom exceptions that spell out to the user what went wrong, so no need to print anything.

Reply:
So far I've been running by the idea that errors should be caught somewhere in the library if they can happen. So far I've been running with QoL as my number 1 priority in terms of making sure that something like the ultrasonic timing out doesn't lead to everything else breaking. There really aren't many situations where an exception can be thrown from the library side, but so far I've been leaning in favor of catching errors and then returning an error code esque value when something goes wrong (like returning 65535 for distance if it times out) Definitely not sure that's the best solution, so lmk what you think

Reply:
Good point, totally agree with QoL being priority. I think we should distinguish between critical and non-critical failures. I would argue that something like the ultrasonic sensor timing out would be non-critical, and shouldn't bring everything to a halt, whereas something like the IMU not being detected during initialization is a critical failure, and should stop everything to alert the user. We may have different opinions on what constitutes critical and non-critical failures, but I do think there are cases when everything needs to stop so the user is aware of the problem.

As for indicating non-critical errors, I'm hesitant about returning special numbers. Since we're targeting people who have little technical experience, I don't think we can't assume they know big powers of 2, and that big powers of 2 are sometimes used to indicate special information. Negative values are arguably a bit better, but that doesn't work for all cases (sometimes -1 is a valid value, for example), and still doesn't clearly convey the information. Maybe we can return something else entirely, like `None` or a string with the error in plain text? Or have some other member function like `get_error_code()` that returns the most recent non-critical failure that occurred? Idk, not sure what's best either.

Reply:
If we're going to return a string as an error code, we might as well raise an exception. I'd be okay changing to -1 as a timeout code for ultrasonic, but part of the reason I had it at 65535 was so that if anybody was trying to drive until they were within a given range, a timeout wouldn't make them stop. I agree with critical errors raising exceptions, but the imu is technically a non-essential function since the drivetrain class doesn't actually need it to operate. That's why I'm caught between just raising an error or just letting it be None and giving a warning

Reply:
maybe have the IMU raise an exception and the drivetrain get_default method catches it if we decide to go down that route?

Reply:
All good points! I guess one other consideration we should make is how frequently we expect these problems to occur. For example, I expect the IMU to fail very rarely (it's soldered directly to the board and tested in production), whereas something like the ultrasonic sensor is much more likely to go wrong (users may forget to plug the sensor in, or wire it up incorrectly, or there's an intermittent connection, or...).

As for how to indicate problems, I don't know if there's a "best" solution here, each seems to have a downside. But after thinking about it some more, I'm personally leaning towards throwing exceptions in most cases. Exceptions explain the problem to the user in plain text, so there's no ambiguity about what the problem is. If it's something that's easily fixed (eg. plug in the ultrasonic cable), then the user can quickly fix it and keep going. If it's something that's not easily fixed (eg. the IMU fails to communicate), the user should really be informed about it. Plus, it's pretty easy for the user to catch any exceptions that come up, and I think it'd be good for students to learn about exceptions at some point.

Truthfully, I don't really have any strong opinions on the best way to indicate problems. Each has their benefits and drawbacks, so I don't think any of them are bad choices. I don't think I have anything more to say on the subject, so I'll leave it up to you!

Reply:
I think that all makes sense, yeah. Going to resolve this comment for now, I'll message in Slack if I have any other questions on how to best improve anything